### ISSUE TITLE
auth.proxy failing for re-created user with same username.

### ISSUE THREAD

**Issue by chandan-groots** [grafana/grafana#19876]
**What happened**:
Not able to authenticate re-created user with auth.proxy

**How to reproduce it (as minimally and precisely as possible)**:
1. Create user with username "**test**"
2. curl -H "X-WEBAUTH-USER: **test**" http://localhost:3000/api/user
3. Successfully able to get response regarding "test" user
4. Delete user "**test**"
4. Re-create user with same username as "**test**"
5. curl -H "X-WEBAUTH-USER: **test**" http://localhost:3000/api/user
6. Getting response as "**User not Found**"

**Environment**:
- Grafana version: Version 6.5.0-pre (commit: 241a0ee, branch: master)
- OS Grafana is installed on: CentOS

**Comment by papagian** [comment]
Thank you for reporting this.
Are you using apache or nginx auth proxy?
How do you create/delete the user? Can you provide the response output from your browser's network tab?

**Comment by chandan-groots** [comment]
Hi papagian,
Apologies for delayed response.
Please find details below:

**>Are you using apache or nginx auth proxy?**
A: I am using Apache auth proxy, but that doesn't matter since the issue is reproducible directly on grafana's port(3000) without using any auth proxy(apache or nginx).

**>How do you create/delete the user?**
A: I created/deleted user using Grafana GUI and I successfully verified that user was indeed created and deleted in SQLite user table.

**>Can you provide the response output from your browser's network tab?**
A: I am using curl to hit the user API, following are details:

Getting info about "test" user:
Command: _curl -H "X-WEBAUTH-USER: test" http://localhost:3000/api/user_
Response: {"id":31,"email":"test","name":"test","login":"test","theme":"","orgId":1,"isGrafanaAdmin":false,"isDisabled":false,"isExternal":true,"authLabels":["OAuth"],"updatedAt":"2019-10-17T16:50:12+05:30","createdAt":"2019-10-17T16:43:38+05:30"}

Getting info about "test" user after deleting and re-creating it.
Command: _curl -H "X-WEBAUTH-USER: test" http://localhost:3000/api/user_
Response: 
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
    <meta name="viewport" content="width=device-width">
    <meta name="theme-color" content="#000">

    <title>Grafana - Error</title>

    <base href="/grafana/" />

    <link rel="stylesheet" href="public/build/grafana.dark.0d9e0fc7f5c3c963471d.css">

    <link rel="icon" type="image/png" href="public/img/fav32.png">
    <link rel="mask-icon" href="public/img/grafana_mask_icon.svg" color="#F05A28">

  </head>

  <body class="theme-dark">
    <div class="main-view">
      <div class="page-container">
        <div class="page-header">
          <div class="page-header__inner">
            <span class="page-header__logo">
              <i class="page-header__icon fa fa-frown-o"></i>
            </span>
            <div class="page-header__info-block">
              <h1 class="page-header__title">
                <a class="text-link" href="login">Grafana</a><span> / Server Error</span><span></span>
              </h1>
              <div class="page-header__sub-title">Sadly something went wrong</div>
            </div>
          </div>
        </div>
      </div>
      <div class="page-container page-body ng-scope" style="padding: 2rem">
        <div class="alert">
          <div class="alert-icon"><i class="fa fa-exclamation-triangle"></i></div>
          <div class="alert-body">
            <div class="alert-title">User not found</div>
          </div>
        </div>
        <br />
        
        <div style="padding: 2rem 0 0">
          <p>Check the Grafana server logs for the detailed error message.</p>
        </div>
      </div>
    </div>
  </body>
</html>

**Comment by papagian** [comment]
Sorry for the delayed reply.
Thank you for reporting this. I was not able to reproduce it without auth proxy but I managed to reproduced it using nginx proxy.
The actual error is [this](https://github.com/grafana/grafana/blob/master/pkg/middleware/auth_proxy.go#L65) and it fails because the `id` parameter passed in `GetSignedUser` is the old row id of the `user` table instead the new one.
So the problem should be that the old record is not deleted also from the [cache](
https://github.com/grafana/grafana/blob/master/pkg/middleware/auth_proxy/auth_proxy.go#L166).

**Comment by papagian** [comment]
@chandan-groots can you verify that you recreated the user in less than one minute after having deleted (which is the period that a key stays in the cache)

**Comment by davidnewhall** [comment]
I'm having this exact same issue on 6.5.1. I deleted a user that was created via auth proxy and now that users gets a 407 return code. Very reproducible with curl as shown by the OP.

```
$ curl -H "X-WEBAUTH-USER: david" http://192.168.3.3:3000/api/user
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
    <meta name="viewport" content="width=device-width">
    <meta name="theme-color" content="#000">
```

```
$ curl -H "X-WEBAUTH-USER: david2" http://192.168.3.3:3000/api/user
{"id":8,"email":"david2","name":"","login":"david2","theme":"","orgId":1,"isGrafanaAdmin":false,"isDisabled":false,"isExternal":true,"authLabels":["OAuth"],"updatedAt":"2019-12-11T15:34:00-08:00","createdAt":"2019-12-11T15:34:00-08:00"}
```

Every user I deleted using the UI is now unable to be used again.

**Comment by cantino** [comment]
We have this issue as well. Running `delete from cache_data` fixes it. I think there is a cache that is not being cleared on user deletion. Is this an easy fix?

**Comment by marefr** [comment]
This should have been fixed by #23979 and include in Grafana v7.1. Closing this for now. Let us know if you believe this is still a problem and we'll consider reopen the issue.
